src/game.rs

```diff
@@ -0,0 +1,340 @@
+    }
+
+    fn input_to_action(act: &char) -> Action {
+        match act {
+            '4' => Action::MoveLeft,
+            '6' => Action::MoveRight,
+            '2' => Action::FallFast,
+            ' ' => Action::Rotate,
+            _ => Action::Fall,
+        }
+    }
+
+    fn make_action(&mut self, act: &Action) {
+        match act {
+            Action::MoveRight => self.move_block(MoveDir::Right),
+            Action::MoveLeft => self.move_block(MoveDir::Left),
+            Action::Fall => self.move_block(MoveDir::Down),
+            Action::Rotate => self.rotate_block(),
+            Action::FallFast => self.fall_fast(),
+        }
+    }
+
+    fn fall_block_down(&mut self) {
+        self.move_block(MoveDir::Down);
+        self.map.print_map();
+    }
+
+    fn rotate_block(&mut self) {
+        match self.moving_block_id {
+            None => (),
+            Some(id) => {
+                if let BlockType::O = self.blocks[id].get_block_type() {
+                    ();
+                } else {
+                    let rotation_coords;
+                    if let BlockType::I = self.blocks[id].get_block_type() {
+                        rotation_coords =
+                            rotate::rotate_i_block(&self.blocks[id].get_blocks_coords());
+                    } else {
+                        let area_coord = self.blocks[id].get_block_min_xy();
+                        let area = self
+                            .map
+                            .get_3x3_clone(area_coord.x as usize, area_coord.y as usize);
+
+                        rotation_coords = rotate::rotate_block(area, id);
+                    }
+                    match self.check_overbounding(&rotation_coords, id) {
+                        CollisionEffect::None => {
+                            self.unplace_block_from_map(id);
+                            self.blocks[id].set_block_coords(rotation_coords);
+                            self.place_block_on_map(id);
+                        }
+                        _ => (),
+                    }
+                }
+            }
+        }
+    }
+
+    fn fall_fast(&mut self) {
+        'fall_loop: loop {
+            match self.moving_block_id {
+                None => (),
+                Some(id) => {
+                    let move_try = self.blocks[id].try_move(&MoveDir::Down);
+                    match self.check_overbounding(&move_try, id) {
+                        CollisionEffect::Stop => {
+                            self.moving_block_id = None;
+                            break 'fall_loop;
+                        }
+                        _ => {
+                            self.move_block(MoveDir::Down);
+                        }
+                    }
+                }
+            }
+        }
+        // wait for long time, so game wont drop down many blocks in row at the same time
+        thread::sleep(Duration::new(0, config::TURN_TIME_NS * 5));
```

Comment:
Jak rozumiem, to jest brzydkie rozwiązenie kwestii długo wciśniętego przycisku. Input powinien wykrywać moment wciśnięcia przycisku i tylko raz wysłać akcję, a nie wysyłać ją cały czas gdy klawisz jest wciśnięty. Podobny problem jest z pozostałymi akcjami. Często wykonują się klika razy z rzędu, gdy gracz wciska przycisk.